tools/src/wyvern/target/corewyvernIL/expression/MethodCall.java

```diff
@@ -69,72 +71,83 @@ public String getMethodName() {
 
 	@Override
 	public ValueType typeCheck(TypeContext ctx) {
-		ValueType ot = objectExpr.typeCheck(ctx);
-		StructuralType st = ot.getStructuralType(ctx);
-
-		List<DeclType> dts = st.findDecls(methodName, ctx);
-		if (dts.isEmpty()) {
+		
+		// Typecheck receiver.
+		ValueType receiver = objectExpr.typeCheck(ctx);
+		StructuralType receiverType = receiver.getStructuralType(ctx);
+
+		// Sanity check: make sure it has declarations.
+		List<DeclType> declarationTypes = receiverType.findDecls(methodName, ctx);
+		if (declarationTypes.isEmpty()) {
 			ToolError.reportError(ErrorMessage.NO_SUCH_METHOD, this, methodName);
 		}
 
-		ArrayList<ValueType> argTypeList = new ArrayList<ValueType>(args.size());
-		for (int i = 0; i < args.size(); ++i) {
-			argTypeList.add(args.get(i).typeCheck(ctx));
-		}
-
+		// Go through all declarations, typechecking against the actual types passed in...
+		List<ValueType> actualArgTypes = getArgTypes(ctx);
+		
+		// ...use this context to do that.
 		TypeContext newCtx = null;
-		for (DeclType dt : dts) {
+		for (DeclType declType : declarationTypes) {
```

Comment:
Is it possible to reuse code between here and getMethodDeclaration?  If so, it might help make this class more maintainable.